File: sdk/lx-trading-go/risk.go
// Package trading provides a unified HFT trading SDK with multi-venue support.
package trading

import (
	"sync"

	"github.com/shopspring/decimal"
)

// RiskManager enforces trading risk limits.
type RiskManager struct {
	config     RiskConfig
	positions  map[string]decimal.Decimal
	dailyPnL   decimal.Decimal
	openOrders map[string]int // symbol -> count
	killSwitch bool

	mu sync.RWMutex
}

// NewRiskManager creates a new risk manager.
func NewRiskManager(config RiskConfig) *RiskManager {
	return &RiskManager{
		config:     config,
		positions:  make(map[string]decimal.Decimal),
		openOrders: make(map[string]int),
	}
}

// IsEnabled returns whether risk management is enabled.
func (r *RiskManager) IsEnabled() bool {
	return r.config.Enabled
}

// IsKilled returns whether the kill switch is active.
func (r *RiskManager) IsKilled() bool {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return r.killSwitch
}

// Kill activates the kill switch, stopping all trading.
func (r *RiskManager) Kill() {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.killSwitch = true
}

// Reset deactivates the kill switch.
func (r *RiskManager) Reset() {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.killSwitch = false
}

// ValidateOrder validates an order against risk limits.
// Returns an error if the order violates any limit.
func (r *RiskManager) ValidateOrder(request OrderRequest) error {
	if !r.config.Enabled {
		return nil
	}

	r.mu.RLock()
	defer r.mu.RUnlock()

	// Check kill switch
	if r.killSwitch {
		return NewRiskError("kill_switch", "active", "inactive")
	}

	// Check order size
	if !r.config.MaxOrderSize.IsZero() && request.Quantity.GreaterThan(r.config.MaxOrderSize) {
		return NewRiskError("order_size", request.Quantity.String(), r.config.MaxOrderSize.String())
	}

	// Check position limit
	pair := ParseTradingPair(request.Symbol)
	if pair.Base != "" {
		current := r.positions[pair.Base]

		var newPosition decimal.Decimal
		if request.Side == SideBuy {
			newPosition = current.Add(request.Quantity)
		} else {
			newPosition = current.Sub(request.Quantity)
		}

		// Asset-specific limit
		if limit, ok := r.config.PositionLimits[pair.Base]; ok {
			if newPosition.Abs().GreaterThan(limit) {
				return NewRiskError(pair.Base+"_position", newPosition.Abs().String(), limit.String())
			}
		}

		// Global position limit
		if !r.config.MaxPositionSize.IsZero() && newPosition.Abs().GreaterThan(r.config.MaxPositionSize) {
			return NewRiskError("position_size", newPosition.Abs().String(), r.config.MaxPositionSize.String())
		}
	}

	// Check open orders count
	count := r.openOrders[request.Symbol]
	if count >= r.config.MaxOpenOrders {
		return NewRiskError("open_orders", itoa(count), itoa(r.config.MaxOpenOrders))
	}

	// Check daily loss
	if !r.config.MaxDailyLoss.IsZero() && r.dailyPnL.IsNegative() {
		if r.dailyPnL.Abs().GreaterThan(r.config.MaxDailyLoss) {
			return NewRiskError("daily_loss", r.dailyPnL.Abs().String(), r.config.MaxDailyLoss.String())
		}
	}

	return nil
}

// UpdatePosition updates position after a trade.
func (r *RiskManager) UpdatePosition(asset string, quantity decimal.Decimal, side Side) {
	r.mu.Lock()
	defer r.mu.Unlock()

	current := r.positions[asset]
	if side == SideBuy {
		r.positions[asset] = current.Add(quantity)
	} else {
		r.positions[asset] = current.Sub(quantity)
	}
}

// UpdatePnL updates daily PnL.
func (r *RiskManager) UpdatePnL(pnl decimal.Decimal) {
	r.mu.Lock()
	defer r.mu.Unlock()

	r.dailyPnL = r.dailyPnL.Add(pnl)

	// Auto kill switch
	if r.config.KillSwitchEnabled && !r.config.MaxDailyLoss.IsZero() {
		if r.dailyPnL.IsNegative() && r.dailyPnL.Abs().GreaterThan(r.config.MaxDailyLoss) {
			r.killSwitch = true
		}
	}
}

// OrderOpened increments open orders count.
func (r *RiskManager) OrderOpened(symbol string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.openOrders[symbol]++
}

// OrderClosed decrements open orders count.
func (r *RiskManager) OrderClosed(symbol string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	if r.openOrders[symbol] > 0 {
		r.openOrders[symbol]--
	}
}

// Position returns current position for an asset.
func (r *RiskManager) Position(asset string) decimal.Decimal {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return r.positions[asset]
}

// Positions returns all positions.
func (r *RiskManager) Positions() map[string]decimal.Decimal {
	r.mu.RLock()
	defer r.mu.RUnlock()

	result := make(map[string]decimal.Decimal, len(r.positions))
	for k, v := range r.positions {
		result[k] = v
	}
	return result
}

// DailyPnL returns current daily PnL.
func (r *RiskManager) DailyPnL() decimal.Decimal {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return r.dailyPnL
}

// ResetDailyPnL resets daily PnL (call at start of trading day).
func (r *RiskManager) ResetDailyPnL() {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.dailyPnL = decimal.Zero
}

// OpenOrderCount returns open orders count for a symbol.
func (r *RiskManager) OpenOrderCount(symbol string) int {
	r.mu.RLock()
	defer r.mu.RUnlock()
	return r.openOrders[symbol]
}

// SetPositionLimit sets position limit for an asset.
func (r *RiskManager) SetPositionLimit(asset string, limit decimal.Decimal) {
	r.mu.Lock()
	defer r.mu.Unlock()
	if r.config.PositionLimits == nil {
		r.config.PositionLimits = make(map[string]decimal.Decimal)
	}
	r.config.PositionLimits[asset] = limit
}

// SetMaxOrderSize sets maximum order size.
func (r *RiskManager) SetMaxOrderSize(max decimal.Decimal) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.config.MaxOrderSize = max
}

// SetMaxDailyLoss sets maximum daily loss.
func (r *RiskManager) SetMaxDailyLoss(max decimal.Decimal) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.config.MaxDailyLoss = max
}

// EnableKillSwitch enables automatic kill switch on max loss.
func (r *RiskManager) EnableKillSwitch(enabled bool) {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.config.KillSwitchEnabled = enabled
}

// ClearPositions clears all position tracking.
func (r *RiskManager) ClearPositions() {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.positions = make(map[string]decimal.Decimal)
}

// ClearOpenOrders clears open orders tracking.
func (r *RiskManager) ClearOpenOrders() {
	r.mu.Lock()
	defer r.mu.Unlock()
	r.openOrders = make(map[string]int)
}

// Summary returns a summary of current risk state.
func (r *RiskManager) Summary() RiskSummary {
	r.mu.RLock()
	defer r.mu.RUnlock()

	totalOpenOrders := 0
	for _, count := range r.openOrders {
		totalOpenOrders += count
	}

	return RiskSummary{
		Enabled:         r.config.Enabled,
		KillSwitch:      r.killSwitch,
		DailyPnL:        r.dailyPnL,
		MaxDailyLoss:    r.config.MaxDailyLoss,
		TotalOpenOrders: totalOpenOrders,
		MaxOpenOrders:   r.config.MaxOpenOrders,
		Positions:       r.Positions(),
	}
}

// RiskSummary provides a snapshot of risk state.
type RiskSummary struct {
	Enabled         bool
	KillSwitch      bool
	DailyPnL        decimal.Decimal
	MaxDailyLoss    decimal.Decimal
	TotalOpenOrders int
	MaxOpenOrders   int
	Positions       map[string]decimal.Decimal
}

File: sdk/lx-trading-cpp/src/risk.cpp
// LX Trading SDK - Risk Manager Implementation

#include <lx/trading/risk.hpp>

namespace lx::trading {

RiskManager::RiskManager(const RiskConfig& config) : config_(config) {}

void RiskManager::validate_order(const OrderRequest& request) {
    if (!config_.enabled) return;

    if (kill_switch_.load(std::memory_order_acquire)) {
        throw RiskError("Kill switch is active");
    }

    // Check order size
    if (config_.max_order_size.is_positive() &&
        request.quantity > config_.max_order_size) {
        throw RiskError(
            "Order size " + request.quantity.to_string() +
            " exceeds max " + config_.max_order_size.to_string());
    }

    // Check position limit
    auto pair = TradingPair::from_symbol(request.symbol);
    if (pair) {
        std::string base(pair->base.data());

        Decimal current;
        {
            std::shared_lock lock(position_mutex_);
            auto it = positions_.find(base);
            if (it != positions_.end()) {
                current = it->second;
            }
        }

        Decimal new_position = (request.side == Side::Buy)
            ? current + request.quantity
            : current - request.quantity;

        // Asset-specific limit
        auto limit_it = config_.position_limits.find(base);
        if (limit_it != config_.position_limits.end()) {
            if (new_position.abs() > limit_it->second) {
                throw RiskError(
                    "Position limit exceeded for " + base + ": " +
                    current.to_string() + " + " + request.quantity.to_string() +
                    " > " + limit_it->second.to_string());
            }
        }

        // Global position limit
        if (config_.max_position_size.is_positive() &&
            new_position.abs() > config_.max_position_size) {
            throw RiskError(
                "Max position size exceeded: " +
                new_position.abs().to_string() + " > " +
                config_.max_position_size.to_string());
        }
    }

    // Check open orders count
    {
        std::shared_lock lock(orders_mutex_);
        auto it = open_orders_.find(request.symbol);
        int count = (it != open_orders_.end()) ? it->second : 0;
        if (count >= config_.max_open_orders) {
            throw RiskError(
                "Max open orders (" + std::to_string(config_.max_open_orders) +
                ") reached for " + request.symbol);
        }
    }

    // Check daily loss
    if (config_.max_daily_loss.is_positive()) {
        std::shared_lock lock(pnl_mutex_);
        if (daily_pnl_.is_negative() &&
            daily_pnl_.abs() > config_.max_daily_loss) {
            throw RiskError(
                "Daily loss limit exceeded: " +
                daily_pnl_.abs().to_string() + " > " +
                config_.max_daily_loss.to_string());
        }
    }
}

void RiskManager::update_position(const std::string& asset, Decimal quantity, Side side) {
    std::unique_lock lock(position_mutex_);
    auto& pos = positions_[asset];
    pos = (side == Side::Buy) ? pos + quantity : pos - quantity;
}

Decimal RiskManager::position(const std::string& asset) const {
    std::shared_lock lock(position_mutex_);
    auto it = positions_.find(asset);
    return (it != positions_.end()) ? it->second : Decimal::zero();
}

std::unordered_map<std::string, Decimal> RiskManager::positions() const {
    std::shared_lock lock(position_mutex_);
    return positions_;
}

void RiskManager::update_pnl(Decimal pnl) {
    std::unique_lock lock(pnl_mutex_);
    daily_pnl_ = daily_pnl_ + pnl;

    // Auto kill switch
    if (config_.kill_switch_enabled &&
        config_.max_daily_loss.is_positive() &&
        daily_pnl_.is_negative() &&
        daily_pnl_.abs() > config_.max_daily_loss) {
        kill_switch_.store(true, std::memory_order_release);
    }
}

Decimal RiskManager::daily_pnl() const {
    std::shared_lock lock(pnl_mutex_);
    return daily_pnl_;
}

void RiskManager::reset_daily_pnl() {
    std::unique_lock lock(pnl_mutex_);
    daily_pnl_ = Decimal::zero();
}

void RiskManager::order_opened(const std::string& symbol) {
    std::unique_lock lock(orders_mutex_);
    open_orders_[symbol]++;
}

void RiskManager::order_closed(const std::string& symbol) {
    std::unique_lock lock(orders_mutex_);
    auto it = open_orders_.find(symbol);
    if (it != open_orders_.end() && it->second > 0) {
        it->second--;
    }
}

int RiskManager::open_orders(const std::string& symbol) const {
    std::shared_lock lock(orders_mutex_);
    auto it = open_orders_.find(symbol);
    return (it != open_orders_.end()) ? it->second : 0;
}

bool RiskManager::check_order_size(Decimal quantity) const noexcept {
    if (!config_.max_order_size.is_positive()) return true;
    return quantity <= config_.max_order_size;
}

bool RiskManager::check_position_limit(
    const std::string& asset, Decimal new_position) const noexcept {
    auto it = config_.position_limits.find(asset);
    if (it != config_.position_limits.end()) {
        if (new_position.abs() > it->second) return false;
    }

    if (config_.max_position_size.is_positive()) {
        if (new_position.abs() > config_.max_position_size) return false;
    }

    return true;
}

bool RiskManager::check_daily_loss() const noexcept {
    if (!config_.max_daily_loss.is_positive()) return true;

    std::shared_lock lock(pnl_mutex_);
    return !daily_pnl_.is_negative() ||
           daily_pnl_.abs() <= config_.max_daily_loss;
}

bool RiskManager::check_open_orders(const std::string& symbol) const noexcept {
    std::shared_lock lock(orders_mutex_);
    auto it = open_orders_.find(symbol);
    int count = (it != open_orders_.end()) ? it->second : 0;
    return count < config_.max_open_orders;
}

}  // namespace lx::trading
